app/lib/storage/encryption.ts

```diff
@@ -0,0 +1,159 @@
+  /** PBFDK2 salt byte length */
+  public saltBytes: number
+
+  constructor(opts: IProtoOptions) {
+    this.algorithm = opts.algorithm || "aes-256-cbc"
+    this.digest = opts.digest || "sha256"
+    this.iterations = opts.iterations || 100000
+    this.ivBytes = opts.ivBytes || 16
+    this.keyBytes = opts.keyBytes || 32
+    this.password = opts.password
+    this.saltBytes = opts.saltBytes || 32
+
+    this.iv = opts.iv || crypto.randomBytes(this.ivBytes)
+    this.salt = opts.salt || crypto.randomBytes(this.saltBytes)
+  }
+
+  /**
+   * Derive the PBKDF2 key for current options and a certain salt.
+   * @param {Buffer} salt
+   * @returns {Buffer}
+   */
+  public getKey(salt: Buffer = this.salt): Buffer {
+    return crypto.pbkdf2Sync(
+      this.password,
+      salt,
+      this.iterations,
+      this.keyBytes,
+      this.digest
+    )
+  }
+}
+
+/**
+ * Crypto buffers serializer function.
+ * @param {Array<Buffer>} buffers
+ * @returns {Buffer}
+ */
+const serialize = (buffers: Array<Buffer>): Buffer => {
+  const parts: Array<Buffer> = []
+
+  buffers.forEach( (part: Buffer) => {
+    const len = Buffer.alloc(4)
+    len.writeUInt32BE(part.length, 0)
+    parts.push(len)
+    parts.push(part)
+  })
+
+  return Buffer.concat(parts)
+}
+
+/**
+ * Crypto buffers deserializer function.
+ * @param {Buffer} data
+ * @returns {Array<Buffer>}
+ */
+const deserialize = (data: Buffer): Array<Buffer> => {
+  const parts: Array<Buffer> = []
+  const l = data.length
+
+  let idx = 0
+  while (idx < l) {
+    const dlen = data.readUInt32BE(idx)
+    idx += 4
+    const start = idx
+    const end = start + dlen
+    const part = data.slice(start, end)
+    parts.push(part)
+    idx += part.length
+  }
+
+  return parts
+}
+
+/**
+ * Applies SHA256 hash function on a key string.
+ * @param {string} key
+ * @returns {Buffer}
+ */
+export const keyHash = (key: string): Buffer => {
+  return crypto.createHash("sha256").update(key).digest()
+}
+
+/**
+ * Encrypts any kind of object into a Buffer.
+ * @param {Buffer} data
+ * @param {Options} opts
+ * @returns {Buffer}
+ */
+export const encrypt = (data: Buffer, opts: Options): Buffer => {
+  const key = opts.getKey()
+  const cipher = crypto.createCipheriv(opts.algorithm, key, opts.iv)
+  const ciphertext = Buffer.concat([cipher.update(data), cipher.final()])
+  const parts = [opts.iv, ciphertext, opts.salt]
+
+  return serialize(parts)
+}
+
+/**
+ * Decrypts any kind of object from a Buffer.
+ * @param {Buffer} data
+ * @param {Options} opts
+ * @returns {any}
+ */
+export const decrypt = (data: Buffer, opts: Options) => {
+  const parts = deserialize(data)
+  const iv = parts[0]
+  const ciphertext = parts[1]
+  const salt = parts[2]
+  const key = opts.getKey(salt)
+
+  const decipher = crypto.createDecipheriv(opts.algorithm, key, iv)
+  const plaintext = decipher.update(ciphertext)
+  const json = Buffer.concat([plaintext, decipher.final()]).toString()
+
+  return JSON.parse(json)
```

Comment:
To make the decrypt function symmetric with the encrypt one, it could just return the `Buffer` and leave the parse logic (`JSON.parse`) to the caller.